### PR TITLE
creating early I/O coin support pull request for feedback/review purposes Address handling+tests added

### PR DESCRIPTION
**Pull request by iocrypto**
## Checklist

- [ ] Prefix PR title with `[WIP]` if necessary.
- [ ] Add tests to cover changes as needed.
- [ ] Update documentation as needed.

### CODE REVIEW

File: src/Iocoin/Address.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#include "../Data.h"
+#include "../PublicKey.h"
+
+#include <array>
+#include <string>
+
+namespace TW::Iocoin {
+
+class Address {
```
Comment (alejandro-isaza) [inline src/Iocoin/Address.h]: Inherit from `Base58Address` most of this is already implemented there.

File: src/Iocoin/Address.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    /// Determines whether a string makes a valid address, and the prefix is
+    /// within the valid set.
+    static bool isValid(const std::string& string, const std::vector<byte>& validPrefixes);
+
+    /// Initializes a  address with a string representation.
+    explicit Address(const std::string& string);
+
+    /// Initializes a  address with a collection of bytes.
+    explicit Address(const std::vector<uint8_t>& data);
```
Comment (alejandro-isaza) [inline src/Iocoin/Address.h]: This is not necessary.

File: src/Iocoin/Address.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+#include <array>
+#include <string>
+
+namespace TW::Iocoin {
+
+class Address : TW::Base58Address<21> {
+  public:
+    /// Number of bytes in an address.
+    //XXXX static const size_t size = 21;
```
Comment (alejandro-isaza) [inline src/Iocoin/Address.h]: Please remove completely. Also below.

Comment (iocrypto) [reply src/Iocoin/Address.h]: removed it

File: src/Iocoin/Address.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+#include <string>
+
+namespace TW::Iocoin {
+
+class Address : public TW::Base58Address<21> {
+  public:
+
+    /// Determines whether a collection of bytes makes a valid  address.
+    template <typename T>
+    static bool isValid(const T& data) {
```
Comment (alejandro-isaza) [inline src/Iocoin/Address.h]: This is the same as `Base58Address` safe to remove.

File: src/Iocoin/Address.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    static bool isValid(const std::string& string, const std::vector<byte>& validPrefixes);
+
+    /// Initializes a  address with a string representation.
+    explicit Address(const std::string& string) : TW::Base58Address<21>(string) {}
+
+    /// Initializes a  address with a public key and a prefix.
+    Address(const PublicKey& publicKey, uint8_t prefix);
+
+    /// Returns a string representation of the address.
+    std::string string() const;
```
Comment (alejandro-isaza) [inline src/Iocoin/Address.h]: This is the same as `Base58Address` safe to remove.

File: src/Iocoin/Address.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    static bool isValid(const T& data) {
+        return data.size() == size;
+    }
+
+    /// Determines whether a string makes a valid address.
+    static bool isValid(const std::string& string);
+
+    /// Determines whether a string makes a valid address, and the prefix is
+    /// within the valid set.
+    static bool isValid(const std::string& string, const std::vector<byte>& validPrefixes);
```
Comment (alejandro-isaza) [inline src/Iocoin/Address.h]: This is the same as `Base58Address` safe to remove.

File: src/Iocoin/Address.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  public:
+
+    /// Determines whether a collection of bytes makes a valid  address.
+    template <typename T>
+    static bool isValid(const T& data) {
+        return data.size() == size;
+    }
+
+    /// Determines whether a string makes a valid address.
+    static bool isValid(const std::string& string);
```
Comment (alejandro-isaza) [inline src/Iocoin/Address.h]: This is the same as `Base58Address` safe to remove.

File: src/Iocoin/Address.cpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+#include <cassert>
+
+using namespace TW::Iocoin;
+
+Address::Address(const PublicKey& publicKey, uint8_t prefix) {
+    if (publicKey.type() != PublicKeyType::secp256k1) {
+        throw std::invalid_argument("Iocoin::Address needs a compressed SECP256k1 public key.");
+    }
+    bytes[0] = prefix;
+    ecdsa_get_pubkeyhash(publicKey.bytes.data(), HASHER_SHA2_RIPEMD, bytes.data() + 1);
```
Comment (hewigovens) [inline src/Iocoin/Address.cpp]: can we use `publicKey.hash(prefix, Hash::sha256ripemd)`?

**Pull request merged**